### ISSUE TITLE
topology issue related to dead neighbors

### ISSUE THREAD

**Issue by tiresias54** [iotaledger/iri#187]
Hey, after a discussion with CfB about topology he suggested to share with core team. 

about topology degradation because of dead neighbor:
the degradation is induced by slow sync caused by a loss of bandwidth related to a huge number of useless requests to dead nodes done by active nodes
i suggested this:
1) periodically check neighbor activity
2) for dead neighbor, drasticly decreases requests frequency until next neighbor activity check 
3) if neighbor is reactivated at next neighbor check, then set request frequency to normal value, else continue to decreased frequency value
(If i understand correctly, this imply to modify IRI code). 

Maybe it's possible to do this with IXI:
1) periodically check neighbor activity
2) for dead neighbor, remove neighbor until next neighbor activity check.
3) at next neighbor activity, re-add the dead neighbor a short time before the check and test if it's still dead or not.

**Comment by alfredch** [comment]
A dead neighbor should be removed from the neighbors list, as simple as that. The questions is: how to qualify a neighbor as dead.

**Comment by tiresias54** [comment]
sure a dead neighbor should be removed, but users let them and that's why the topology is sometimes bad.
A dead neighbor could be a neighbor which has "numberOfAllTransactions" not evolving on a chosen time step.

**Comment by Sharky444** [comment]
A dead neighbor should not be simply removed. Maybe he experienced some technical difficulty that will be solved 24 hours later.

**Comment by alfredch** [comment]
It is currently not the responsibility of IRI core to manage the physical topology. I agree that temporary neighbors should also be supported. This could be achieved, for instance, by re-adding a neighbor periodically via the addNeighbor API call (every 1 hour, or 24 hours, whatever). If it is not responding or again too slow, then it will be removed automatically after a relatively short scan period (5 minutes).

**Comment by tiresias54** [comment]
Indeed, maybe in IXI? Anyway i'm not the dude to code this as i'm a physicist.

**Comment by paulhandy** [comment]
like this? https://github.com/bluedigits/nstats

**Comment by Sharky444** [comment]
No matter where, but some part of IOTA needs to automatically get rid of non existing neighbours after 24h hours. This would immensely improve the network quality.

**Comment by bluedigits** [comment]
We already have this. Imho open issues are:
- How could a full node manager be notified of a removed neighbor?
- How and when can neighbors be added again (automatically)?

**Comment by Sharky444** [comment]
The node should periodinally output info about active neighbours and their quality. Not as an extension, but as basic functionality.

**Comment by footloosejava** [comment]
Good idea.

**Comment by tiresias54** [comment]
as Andy pointed, what happens if everyone use the module?

**Comment by bluedigits** [comment]
I've integrated automatic neighbor re-adding.

**Comment by eukaryote31** [comment]
Added/removed neighbors should also be persisted to disk. I think we should remove the neighbor list form the config and put it in its own file; that way, it makes updating the neighbors list more consistent and organized.
